utils/yamlConfig.js

```diff
@@ -0,0 +1,146 @@
+
+/*
+  This program and the accompanying materials are
+  made available under the terms of the Eclipse Public License v2.0 which accompanies
+  this distribution, and is available at https://www.eclipse.org/legal/epl-v20.html
+  
+  SPDX-License-Identifier: EPL-2.0
+  
+  Copyright Contributors to the Zowe Project.
+*/
+
+const fs = require('fs');
+const path = require('path');
+const _ = require('lodash');
+const YAWN = require('yawn-yaml/cjs');
+const mergeUtils = require('./mergeUtils');
+
+function encodeKey(key) {
+  return key.replace(/[^a-zA-Z0-9]/g, char => '_x' + char.charCodeAt(0).toString(16));
+}
+
+function flattenEnvObject(obj, prefix) {
+  const result = {};
+  const path = prefix;
+  flattenEnvObjectInternal(obj, path, result);
+  return result;
+
+  function flattenEnvObjectInternal(obj, path, result) {
+    for (const key of Object.keys(obj)) {
+      const encodedKey = encodeKey(key);
+      const newPath = path ? `${path}_${encodedKey}` : encodedKey;
+      const val = obj[key];
+      if (typeof val === 'object') {
+        flattenEnvObjectInternal(val, newPath, result);
+      } else if (Array.isArray(val)) {
+        result[newPath] = val.join(',') + ',';
+      } else {
+        result[newPath] = String(val);
+      }
+    }
+  }
+}
+
+function convertConfigToEnvObj(config) {
+  return flattenEnvObject(config, 'ZWED');
+}
+
+function convertConfigToEnvSource(config) {
+  const envObj = convertConfigToEnvObj(config);
+  return Object.keys(envObj).map(key => `export ${key}="${envObj[key]}"`).join('\n');
+}
+
+function getHaInstanceId() {
+  return process.env['ZWELS_HA_INSTANCE_ID'];
+}
+
+function getInstanceDir() {
+  return process.env['INSTANCE_DIR'] || '~/.zowe';
+}
+
+function omitCommonConfigKeys(config) {
+  const commonConfigKeys = [
+    'certificate',
+    'crossMemoryServerName',
+    'enabled',
+    'launcher',
+    'port',
+    'tls'
+  ];
+  return _.omit(config, commonConfigKeys);
+}
+
+function getComponentConfig(yawn, component, haInstanceId) {
+  const componentLevelConfig = _.get(yawn.json, ['components', component]);
+  const instanceLevelConfig = _.get(yawn.json, ['haInstances', haInstanceId, 'components', component]);
+  const config = mergeUtils.deepAssign(componentLevelConfig, instanceLevelConfig ? instanceLevelConfig : {});
+  return omitCommonConfigKeys(config);
+}
+
+function getYamlConfig(yawn, haInstanceId) {
+  const appServerConfig = getComponentConfig(yawn, 'app-server', haInstanceId);
+  const zssConfig = getComponentConfig(yawn, 'zss', haInstanceId);
```

Comment:
Probably this should also be abstracted. Maybe in the future there will be a different agent other than zss, and we'll want to merge app-server config with another component config.
What if this took an array rather than just 2 components to merge?